reps/2024-05-21-kuberay-authentication.md

```diff
@@ -0,0 +1,159 @@
+This includes enforcing authentication for both the dashboard server and the interactive client server. Securing access from internal clients like the Raylet to the GCS server is not addressed in this proposal
+as network policies are typically sufficient to protect this communication.
+
+## Sidecar Authentication Proxy
+
+To enforce authenticated access to specific Ray head ports, the KubeRay operator will deploy a sidecar container alongside the Ray head pod.
+This sidecar will function as a reverse proxy, validating authentication tokens for requests to the dashboard port (8265) and the interactive client port (10001).
+Traffic to other ports will pass through the sidecar proxy without requiring authentication.
+
+The KubeRay operator will be changed in the following ways:
+1. The Ray head container will bind to localhost only
+2. A reverse proxy sidecar will run alongside the Ray head container
+3. A ServiceAccount is created per RayCluster resource
+
+The sidecar ensures only requests with the correct authorization tokens can access the Ray head. More details on how tokens are authenticated below.
+For the MVP, the sidecar will be implemented using Envoy configured with an [External Authorizer](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_authz_filter).
+
+## Access Control with TokenReview API
+
+The sidecar authenticator will use the TokenReview API to authenticate tokens passed into the authorization headers of each request.
+Tokens can be created from Kubernetes ServiceAccounts or an external identity provider as long as the Kubernetes API Server is configured with an external authorizer.
+Tokens from Kubernetes Service Accounts are required to specify an audience. For now the placeholder audience is "ray.io".
+Each RayCluster will be provisioned a default ServiceAccount that KubeRay Operator will use when authenticating to the Ray head (specifically for RayJob and RayService).
+Users can use the default ServiceAccount in the absence of external identity providers.
+
+Example of a TokenReview request that contains a token:
+```
+apiVersion: authentication.k8s.io/v1
+kind: TokenReview
+spec:
+  token: <your_token_here>
+  audiences:
+  - "ray.io"
+```
+
+Example of a TokenReview response with user information about the token:
+```
+apiVersion: authentication.k8s.io/v1
+kind: TokenReview
+status:
+  authenticated: true
+  user:
+    username: system:serviceaccount:default:ray-cluster
+    uid: <some_uid>
+    groups: ["system:serviceaccounts", "system:serviceaccounts:default"]
+  audiences:
+  - "ray.io"
+```
+
+## User to Ray Head Authentication
+
+When submitting jobs using the Ray job CLI:
+```
+export RAY_JOB_HEADERS="Authorization: Bearer $(kubectl create token ray-cluster)"
+OR
+export RAY_JOB_HEADERS="Authorization: Bearer $(gcloud auth print-access-token)" # example external identity provider
```

Comment:
I was surprised that this would work, but it does seem like TokenReview will understand external tokens by calling the Webhook Token Authenticator, if one is configured.

We'll need to think about how to make this mesh with the audiences from above.  GCP (and presumably AWS, Azure, etc) access tokens don't carry scopes, so they won't pass the scope check by default.

Reply:
> I was surprised that this would work, but it does seem like TokenReview will understand external tokens by calling the Webhook Token Authenticator, if one is configured.

I was also pleasantly surprised as well :) 

> We'll need to think about how to make this mesh with the audiences from above. GCP (and presumably AWS, Azure, etc) access tokens don't carry scopes, so they won't pass the scope check by default.

From my testing w/ GKE, the returned audience is something like:
```
audiences:
- https://container.googleapis.com/v1/projects/<project>/locations/us-east4-c/clusters/<cluster>
```

Perhaps we need to specify a list of allowed audiences in the RayCluster if we are required to verify the audience.

Reply:
OK, yes, that seems like it would work.